panic: pebble: flush-to key (NRK4VxF) < flushed key (NijFwKPxZE)

Reproducible when running the metamorphic test with the standard options `LevelOptions.TargetFileSize = 1`:

```
panic: pebble: flush-to key (NRK4VxF) < flushed key (NijFwKPxZE)

goroutine 10 [running]:
github.com/cockroachdb/pebble/internal/rangedel.(*Fragmenter).FlushTo(0xc00000c2d0, 0xc0001d3ce0, 0x7, 0x8)
	/Users/pmattis/Development/go/src/github.com/cockroachdb/pebble/internal/rangedel/fragmenter.go:291 +0x79e
github.com/cockroachdb/pebble.(*compactionIter).Tombstones(0xc00000c1e0, 0xc0001d3ce0, 0x7, 0x8, 0x7, 0xc0001d3ce0, 0x0)
	/Users/pmattis/Development/go/src/github.com/cockroachdb/pebble/compaction_iter.go:485 +0x11f
github.com/cockroachdb/pebble.(*DB).runCompaction.func4(0xc0001d3580, 0x7, 0x10, 0x2, 0x8, 0xc0001d3710)
	/Users/pmattis/Development/go/src/github.com/cockroachdb/pebble/compaction.go:1038 +0xff
github.com/cockroachdb/pebble.(*DB).runCompaction(0xc0000bea00, 0x19, 0xc000244380, 0x1480d20, 0xc0000903c0, 0x5, 0x8, 0xc000104000, 0x15, 0x0, ...)
	/Users/pmattis/Development/go/src/github.com/cockroachdb/pebble/compaction.go:1116 +0xcc0
github.com/cockroachdb/pebble.(*DB).compact1(0xc0000bea00, 0x0, 0x0)
	/Users/pmattis/Development/go/src/github.com/cockroachdb/pebble/compaction.go:875 +0x519
github.com/cockroachdb/pebble.(*DB).compact(0xc0000bea00)
	/Users/pmattis/Development/go/src/github.com/cockroachdb/pebble/compaction.go:824 +0x71
created by github.com/cockroachdb/pebble.(*DB).maybeScheduleCompaction
	/Users/pmattis/Development/go/src/github.com/cockroachdb/pebble/compaction.go:817 +0xab
```

--- a/compaction_iter.go
+++ b/compaction_iter.go
@@ -189,11 +189,6 @@ func (i *compactionIter) First() (*InternalKey, []byte) {
 	i.iterKey, i.iterValue = i.iter.First()
 	if i.iterKey != nil {
 		i.curSnapshotIdx, i.curSnapshotSeqNum = snapshotIndex(i.iterKey.SeqNum(), i.snapshots)
-		if i.iterKey.Kind() == InternalKeyKindRangeDelete {
-			// Range tombstones are always added to the fragmenter. They are
-			// processed into stripes after fragmentation.
-			i.rangeDelFrag.Add(i.cloneKey(*i.iterKey), i.iterValue)
-		}
 	}
 	return i.Next()
 }
@@ -211,6 +206,11 @@ func (i *compactionIter) Next() (*InternalKey, []byte) {
 	i.valid = false
 	for i.iterKey != nil {
 		i.key = *i.iterKey
+		if i.key.Kind() == InternalKeyKindRangeDelete {
+			// Range tombstones are always added to the fragmenter. They are
+			// processed into stripes after fragmentation.
+			i.rangeDelFrag.Add(i.cloneKey(i.key), i.iterValue)
+		}
 		if i.rangeDelFrag.Deleted(i.key, i.curSnapshotSeqNum) {
 			i.saveKey()
 			i.skipStripe()
@@ -300,15 +300,7 @@ func (i *compactionIter) skipStripe() {
 
 func (i *compactionIter) iterNext() bool {
 	i.iterKey, i.iterValue = i.iter.Next()
-	if i.iterKey == nil {
-		return false
-	}
-	if i.iterKey.Kind() == InternalKeyKindRangeDelete {
-		// Range tombstones are always added to the fragmenter. They are processed
-		// into stripes after fragmentation.
-		i.rangeDelFrag.Add(i.cloneKey(*i.iterKey), i.iterValue)
-	}
-	return true
+	return i.iterKey != nil
 }
 
 func (i *compactionIter) nextInStripe() bool {
@@ -318,10 +310,22 @@ func (i *compactionIter) nextInStripe() bool {
 	key := i.iterKey
 	if i.cmp(i.key.UserKey, key.UserKey) != 0 {
 		i.curSnapshotIdx, i.curSnapshotSeqNum = snapshotIndex(key.SeqNum(), i.snapshots)
+		// If i.iterKey is a range tombstone, it will be added to the fragmenter in
+		// the next call to compactionIter.Next(). This prevents the tombstone from
+		// being emitted to the previous sstable if it couldn't possibly overlap
+		// keys in that sstable.
 		return false
 	}
 	switch key.Kind() {
 	case InternalKeyKindRangeDelete:
+		if key.Kind() == InternalKeyKindRangeDelete {
+			// Range tombstones are always added to the fragmenter. They are
+			// processed into stripes after fragmentation. Note that we only add
+			// range tombstones to the fragmenter when they have the same start key
+			// as i.key. Adding tombstones earlier can violate a rangedel.Fragmenter
+			// invariant and lead to too many tombstones being added to an sstable.
+			i.rangeDelFrag.Add(i.cloneKey(*key), i.iterValue)
+		}
 		return true
 	case InternalKeyKindInvalid:
 		i.curSnapshotIdx, i.curSnapshotSeqNum = snapshotIndex(key.SeqNum(), i.snapshots)

--- a/testdata/compaction_iter
+++ b/testdata/compaction_iter
@@ -905,3 +905,20 @@ next
 ----
 d#2,0:a
 .
+
+define
+a.MERGE.2:a
+b.RANGEDEL.1:c
+----
+
+iter
+first
+tombstones a
+next
+tombstones
+----
+a#2,2:a
+.
+.
+b-c#1
+.

